### ISSUE TITLE
JS Optimization breaks the application, Dev mode works fine

### ISSUE THREAD

**Issue by savanvadalia** [ionic-team/ionic-app-scripts#1194]
#### Short description of the problem:

Unable to use a 3rd party lib [json2typescript ](https://www.npmjs.com/package/json2typescript) in production mode `ionic build --prod`. When running the application in Dev mode it works fine (No issues with `ionic serve`)

#### What behavior are you expecting?
I should be able to use this 3rd party library [json2typescript ](https://www.npmjs.com/package/json2typescript) in production mode to take benefit of all the optimizations.

#### Steps to reproduce:

1) Upon digging further I have found that the issue is with `--optimizejs` part of build.
2) I am able to successfully run a sample blank ionic application which consumes the lib [json2typescript ](https://www.npmjs.com/package/json2typescript) using this build option 
`ionic build --aot --minifyjs --minifycss`
3) As soon as I add `--optimizejs` option it fails when the application loads
`ionic build --aot --minifyjs --minifycss --optimizejs`
4) Here is the sample repo for the blank ionic application
https://github.com/savanvadalia/IonicJson2TypescriptError
5) Please download the repo and execute the app in dev and production mode to see the error
6) All the code which consumes the 3rd party lib is in the [home.ts](
https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/pages/home/home.ts
)

**Which @ionic/app-scripts version are you using?**
```
cli packages: (C:\Users\Laptop-Savan\AppData\Roaming\npm\node_modules)

    @ionic/cli-utils  : 1.9.2
    ionic (Ionic CLI) : 3.9.2

local packages:

    @ionic/app-scripts : 2.1.3
    Ionic Framework    : ionic-angular 3.6.0

System:

    Node : v6.9.1
    npm  : 3.10.8
    OS   : Windows 10
```

**Comment by janpio** [comment]
Do you get any errors when building?
What is the error you are getting when executing the code?
What line exactly is throwing the error?
What exact steps did you take to add the lib to your project? (Your repo is only one commit, not the individual steps taken to add the functionality to the starter)
Can the code example be trimmed down even more and still trigger the problem?

**Comment by savanvadalia** [comment]
Do you get any errors when building?
 -> No
What is the error you are getting when executing the code?
-> Please see screenshot at the end
What line exactly is throwing the error?
-> `Line 53:   in home.ts file  this.country = jsonConvert.deserialize(jsonObject, Country);`
What exact steps did you take to add the lib to your project? (Your repo is only one commit, not the 
individual steps taken to add the functionality to the starter)
-> `npm install json2typescript --save` and then just use the lib functions by importing the lib as shown in [home.ts](https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/pages/home/home.ts)

Can the code example be trimmed down even more and still trigger the problem?
-> Code example is already a very trimmed version of a blank ionic starter app having just one page and it calls one function in the homepage constructor to trigger the lib usage

------------------Error Screenshot------------------
Please view error screenshot by clicking the image for better resolution
![image](https://user-images.githubusercontent.com/17091755/29871405-f9127f52-8dce-11e7-996b-21dc8dc1c2be.png)

**Comment by janpio** [comment]
Thanks.

Can the object be simpler?
https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/pages/home/home.ts#L20-L40

Can the target model be simpler?
https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/model/country.ts

Can the object be defined directly or is the parsing relevant?
https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/pages/home/home.ts#L43

Do these options make a difference?
https://github.com/savanvadalia/IonicJson2TypescriptError/blob/master/src/pages/home/home.ts#L45-L49

**Comment by savanvadalia** [comment]
Can the object be simpler?
-- Yes, have made it simpler and committed the code.

Can the target model be simpler?
-- Yes, have made it simpler and committed the code.

Can the object be defined directly or is the parsing relevant?
-- Yes, removed the json parsing and defined the object directly.

Do these options make a difference?
-- No, that won't make any difference.

**Comment by janpio** [comment]
Ok, that is a pretty simple repro now.

I think it removes the annotations in the optimize step which are required by that library. 
Alternatively it renamed the class to optimize, but of course didn't change the string reference to it in the code or annotation.
There is probably a way to tell whatever library is doing the optimization (webpack?) not to do that.

Can you maybe search for the class in the optimized bundle/file and see what became of it after optimization?

**Comment by savanvadalia** [comment]
I am assuming when you say annotations you are referring to the decorators, which in this case are `JsonObject` and `JsonProperty`?

@danbucholtz suggested me a workaround until he can look into this issue, but that doesn't work either. He asked me to override build properties in the `package.json` config section like below.
```
  "config": {
    "ionic_purge_decorators": false,
    "ionic_manual_treeshaking": false
  },
```
He advised me to use combination of both and none the work yet and give different errors.

1)
 ```
  "config": {
    "ionic_purge_decorators": false,
    "ionic_manual_treeshaking": false
  },
```
Same error as above

2)
 ```
  "config": {
    "ionic_purge_decorators": false,
    "ionic_manual_treeshaking": true
  },
Different error with this option
```
![image](https://user-images.githubusercontent.com/17091755/29958256-1e7b96c2-8f35-11e7-8da0-096310b23fd0.png)

3)
```
  "config": {
    "ionic_purge_decorators": true,
    "ionic_manual_treeshaking": false
  },
```
Different error with this option
![image](https://user-images.githubusercontent.com/17091755/29958308-72826958-8f35-11e7-88d5-36ed4d127ac1.png)

I am finding hard to work with an optimized bundle/file not sure I understand the code after the optimization process. I wish if you or someone else can help here.

**Comment by janpio** [comment]
Yes, I was referring to these. And `"ionic_purge_decorators": false` seems to be to do exactly what I suggested as well.

Sorry, I am out of ideas. Same with the error messages you are getting.

**Comment by savanvadalia** [comment]
No probs, thanks for giving try.

**Comment by janpio** [comment]
(You could post at https://forum.ionicframework.com/ or post a link to an existing topic)

**Comment by savanvadalia** [comment]
Thanks posted a link in the forum too

**Comment by janpio** [comment]
(Can you post the topic link here so me and all others reading can make the connection?)

**Comment by rstraub** [comment]
Any development on this issue? I am experiencing a similar issue when importing the `web-animations-js` library in my application. It works great in development but with the --prod flag it doesnt work.

**Comment by andreas-aeschlimann** [comment]
Make sure you do not have any circular dependencies. We did not have any issues as soon as we made sure that they were gone.

**Comment by leonardosalles** [comment]
I am facing a problem with --optimize-js flag when --prod is used using version 3.1.9. Solved with  downgrade to 3.1.6. Everything works fine right now.
